module/Activity/src/Activity/Controller/ActivityController.php

```diff
@@ -131,223 +168,246 @@ protected function prepareSignupForm($fields, & $activitySession, & $externalSes
     public function createAction()
     {
         $activityService = $this->getServiceLocator()->get('activity_service_activity');
-        $form = $activityService->getForm();
-        if ($this->getRequest()->isPost()) {
-            $postData = $this->getRequest()->getPost();
-            $form->setData($postData);
+        $form = $activityService->getActivityForm();
+        $request = $this->getRequest();
 
-            if ($form->isValid()) {
-                $activityService->createActivity(
-                    $form->getData(\Zend\Form\FormInterface::VALUES_AS_ARRAY),
-                    $postData['language_dutch'],
-                    $postData['language_english']
-                );
+        if ($request->isPost()) {
+            if ($activityService->createActivity($request->getPost())) {
                 $view = new ViewModel();
                 $view->setTemplate('activity/activity/createSuccess.phtml');
+
                 return $view;
             }
         }
 
-        return ['form' => $form];
+        $translator = $this->getServiceLocator()->get('translator');
+
+        return [
+            'form' => $form,
+            'action' => $translator->translate('Create Activity'),
+            'allowSignupList' => true,
+        ];
     }
 
     /**
      * Signup for a activity.
      */
     public function signupAction()
```

Comment:
The signupAction has lots of similarities with the externalSignupAction. Both functions are quite complex as well. Do you think it is possible to extract some common functionalities to seperate method(s)?